src/ripple/app/tx/impl/DID.cpp

```diff
@@ -0,0 +1,215 @@
+*/
+//==============================================================================
+
+#include <ripple/app/tx/impl/DID.h>
+
+#include <ripple/basics/Log.h>
+#include <ripple/ledger/ApplyView.h>
+#include <ripple/ledger/View.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Indexes.h>
+#include <ripple/protocol/TxFlags.h>
+#include <ripple/protocol/st.h>
+
+namespace ripple {
+
+/*
+    DID
+    ======
+
+    TODO: add docs here
+*/
+
+//------------------------------------------------------------------------------
+
+NotTEC
+DIDSet::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureDID))
+        return temDISABLED;
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+        return temINVALID_FLAG;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (!ctx.tx.isFieldPresent(sfURI) && !ctx.tx.isFieldPresent(sfData))
+        return temEMPTY_DID;
+
+    if (ctx.tx.isFieldPresent(sfURI) && ctx.tx[sfURI].empty() &&
+        ctx.tx.isFieldPresent(sfData) && ctx.tx[sfData].empty())
+        return temEMPTY_DID;
+
+    return preflight2(ctx);
+}
+
+TER
+addSLE(
+    ApplyContext& ctx,
+    std::shared_ptr<SLE> const& sle,
+    AccountID const& owner)
+{
+    auto const sleAccount = ctx.view().peek(keylet::account(owner));
+    if (!sleAccount)
+        return tefINTERNAL;
+
+    // Check reserve availability for new object creation
+    {
+        auto const balance = STAmount((*sleAccount)[sfBalance]).xrp();
+        auto const reserve =
+            ctx.view().fees().accountReserve((*sleAccount)[sfOwnerCount] + 1);
+
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    // Add ledger object to ledger
+    ctx.view().insert(sle);
+
+    // Add ledger object to owner's page
+    {
+        auto page = ctx.view().dirInsert(
+            keylet::ownerDir(owner), sle->key(), describeOwnerDir(owner));
+        if (!page)
+            return tecDIR_FULL;
+        (*sle)[sfOwnerNode] = *page;
+    }
+    adjustOwnerCount(ctx.view(), sleAccount, 1, ctx.journal);
+    ctx.view().update(sleAccount);
+
+    return tesSUCCESS;
+}
+
+TER
+DIDSet::doApply()
+{
+    // Edit ledger object if it already exists
+    Keylet const didKeylet = keylet::did(account_);
+    if (auto const sleDID = ctx_.view().peek(didKeylet))
+    {
+        if (auto const uri = ctx_.tx[~sfURI])
+        {
+            if (uri->empty())
+            {
+                sleDID->makeFieldAbsent(sfURI);
+            }
+            else
+            {
+                (*sleDID)[sfURI] = *uri;
+            }
+        }
+        if (auto const data = ctx_.tx[~sfData])
+        {
+            if (data->empty())
+            {
+                sleDID->makeFieldAbsent(sfData);
+            }
+            else
+            {
+                (*sleDID)[sfData] = *data;
+            }
+        }
+        if (!sleDID->isFieldPresent(sfURI) && !sleDID->isFieldPresent(sfData))
```

Comment:
This bit could be done in `preclaim`; not sure it makes sense to do it, just pointing it out.

Reply:
It seems overly complicated to add that to `preclaim`.

Reply:
Agreed. And any performance gains are likely to be overshadowed by the extra work to load/manipulate the objects.